File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/factory/transportmethod/TransportFactory.java
package com.bigboxcode.designpattern.factory.transportmethod;

public abstract class TransportFactory {
    public void operateTransport(String name) {
        Transport transport = getTransport(name);

        transport.start();
        transport.stop();
    }

    public void repairTransport(String name) {
        Transport transport = getTransport(name);

        transport.repair();
    }

    public abstract Transport getTransport(String name);
}

File: design-patterns/ts/factory/transport-factory-method/transport-factory.ts
// transport-factory.ts

import Transport from "./transport";

abstract class TransportFactory {
    operateTransport(name: string): void {
        const transport = this.getTransport(name);

        transport?.start();
        transport?.stop();
    }

    repairTransport(name: string): void {
        const transport = this.getTransport(name);

        transport?.repair();
    }

    abstract getTransport(name: string): (Transport | null);
}

export default TransportFactory;